File: 1352-maximum-profit-in-job-scheduling/1352-maximum-profit-in-job-scheduling.cpp
class Solution {
    static bool comparepair(pair<pair<int, int>, int>& a,
                     pair<pair<int, int>, int>& b) {
        return a.first.first < b.first.first;
    }
    bool consistent(pair<int, int> a, pair<int, int> b) {
        return !(a.first < b.second && b.first < a.second);
    }
    int nextConsistent(vector<pair<pair<int, int>, int>>& intervals, int l,
                       int r) {
        int index = -1;
        int prev = l - 1;
        int mid;
        while (l <= r) {
            mid = (l + r) / 2;
            if (consistent(intervals[prev].first, intervals[mid].first)) {
                index = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return index;
    }
    int jobSchedul(vector<pair<pair<int, int>, int>>& intervals, int n, int i,
                   vector<int>& dp) {
        if (i >= n) {
            return 0;
        }

        if (dp[i] != -1) {
            return dp[i];
        }

        int take = intervals[i].second, notTake = 0;
        int nextConsistentinexd = nextConsistent(intervals, i + 1, n - 1);

        if (nextConsistentinexd != -1) {
            take += jobSchedul(intervals, n, nextConsistentinexd, dp);
        }
        notTake = jobSchedul(intervals, n, i + 1, dp);

        return dp[i] = max(take, notTake);
    }

public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime,
                      vector<int>& profit) {
        int n = startTime.size();
        vector<pair<pair<int, int>, int>> intervals(n);

        for (int i = 0; i < n; i++) {
            intervals[i] = {{startTime[i], endTime[i]}, profit[i]};
        }

        sort(intervals.begin(), intervals.end(), comparepair);
        vector<int> dp(n, -1);
        jobSchedul(intervals, n, 0, dp);

        for (int i = 0; i < n; i++)
            cout << dp[i] << " ";

        return dp[0];
    }
};

File: 1352-maximum-profit-in-job-scheduling/1352-maximum-profit-in-job-scheduling.java
class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int len = startTime.length;
        int[][]activities = new int[len][3];
        
        for (int i =0 ; i < len; i++){
            activities[i][0] = startTime[i];
            activities[i][1] = endTime[i];
            activities[i][2] = profit[i];
        }
        
        return solve(activities);
    }

    public boolean nonOverlap(int[]first, int[]second){
        return second[0] >= first[1];
    }
    public int nextNonOverlapActivity(int[][]activities, int index){
        int l = index + 1;
        int r = activities.length - 1;

        int nextIndex = -1;
        while (l <= r){
            int mid = (l + r) / 2;

            if(nonOverlap(activities[index], activities[mid])){
                nextIndex = mid;
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }

        return nextIndex;
    }
    public int maxNonOverlapActivities(int[][]activities, int[]dp, int index, int n){
        if (index >= n){
            return 0;
        }
        if(dp[index] != -1){
            return dp[index];
        }

        int take = activities[index][2];
        int nextNonOverlapIndex = nextNonOverlapActivity(activities, index);
        if(nextNonOverlapIndex != -1){
            take += maxNonOverlapActivities(activities, dp, nextNonOverlapIndex, n);
        }

        int notTake = maxNonOverlapActivities(activities, dp, index + 1, n);

        return dp[index] = Math.max(take, notTake);
    }
    public int solve(int[][]activities){
        Arrays.sort(activities, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return Integer.compare(o1[0], o2[0]);
            }
        });

        int activitiesLen= activities.length;
        int[]dp = new int[activitiesLen];
        Arrays.fill(dp, -1);

        return maxNonOverlapActivities(activities, dp, 0, activitiesLen);
    }
}